Allow starting agents from remote webdriver servers

cowboyd opened the pull request:
Motivation
-----------

Currently, it is assumed that BigTest will manage your webdriver server in-process, both starting it and stopping it on demand. In fact, for things like geckodriver which only allows a single browser to be run per-server, there is a server process per web agent.

This causes a bunch of issues because it means that the bigtest user has to manage their own browser binaries since their server must ultimately start and stop browser processes. This change allows bigtest to create agents directly off of an existing webdriver server. Among other things, this opens the door to running browsers completely off-container in things like BrowserStack or Saas labs, but it could also be running off of a server started in the container before the test run.

Approach
----------
To do this, the driver options have been expanded to account for both "remote" and "local" options. The type of driver is used to resolve the url of the driver server before ultimately using that url to connect and disconnect new web drivers.

cowboyd committed:
Allow starting agents from remote webdriver servers

Currently, it is assumed that BigTest will manage your webdriver
server in-process, both starting it and stopping it on demand. In
fact, for things like geckodriver which only allows a single browser
to be run per-server, there is a server process per web agent.

This causes a bunch of issues because it means that the bigtest user
has to manage their own browser binaries since their server must
ultimately start and stop browser processes. This change allows
bigtest to create agents directly off of an existing webdriver
server. Among other things, this opens the door to running browsers
completely off-container in things like BrowserStack or Saas labs, but
it could also be running off of a server started in the container
before the test run.

To do this, the driver options have been expanded to account for both
"remote" and "local" options. The type of driver is used to resolve
the url of the driver server before ultimately using that url to
connect and disconnect new web drivers.

dagda1 commented on packages/webdriver/src/web-driver.ts:
```diff
@@ -34,7 +40,7 @@ export function* connect(driver: WebDriver, options: Options): Operation<void> {
   let capabilities = new Atom(Capabilities);
 
-  if (options.headless) {
+  if (options.headless != null && options.headless) {
```
is that an intentional `!=` instead of `!==`?

would `!!options.headless` work instead?

I would have the [eqeqeq](https://eslint.org/docs/rules/eqeqeq) rule set in eslint for strict equals

cowboyd replied:
I had done this originally so that specifying the boolean option wasn't _required_ so only do the check if it is not undefined, but I guess if it's undefined it will have the same effect as being false.

dagda1 commented on packages/webdriver/src/web-driver.ts:
```diff
@@ -44,37 +50,61 @@ export function* connect(driver: WebDriver, options: Options): Operation<void> {
+  headless?: boolean;
+};
+
+export type Options = LocalOptions | RemoteOptions;
```
Good use of a discriminated union or adt!

dagda1 reviewed (approved):
This is really great.

cowboyd committed:
removed useless defined check

cowboyd committed:
update to timeout-free untilURLAvailable()

cowboyd committed:
make check against `remote`

This will make the local webdriver the default and thus make
`bigtest.json` backwards compatible
